File: ChurnReduction.R
rm(list=ls())
setwd("/Users/divyasharma/Documents/Edwisor/Project 1/")
getwd()
churn_train=read.csv("Train_data.csv", header = T)
churn_test=read.csv("Test_data.csv", header = T)

#Merge two dataframe to perform data pre-processing
data=rbind(churn_train,churn_test)


#load all the libraries
x = c("ggplot2", "corrgram","randomForest",  "C50","caret","DataCombine")

install.packages(x)
rm(x)
#Missing value analysis:

tot=data.frame(sapply(data,function(x){sum(is.na(x))}))
tot
#There are no missing values in our dataset

##Data Manupulation; convert string categories into factor numeric
for(i in 1:ncol(data)){
  
  if(class(data[,i]) == 'factor'){
    
    data[,i] = factor(data[,i], labels=(1:length(levels(factor(data[,i])))))
  }
}

##Outlier Analysis
cnames = c("number.vmail.messages","total.day.minutes",
           "total.day.calls",
           "total.day.charge","total.eve.minutes",
           "total.eve.calls","total.eve.charge","total.night.minutes",
           "total.night.calls","total.night.charge","total.intl.minutes",
           "total.intl.calls","total.intl.charge","number.customer.service.calls")
 

 
for (i in cnames)
{
boxplot(data[,i] ~ Churn, data = data,
        notch = TRUE, col = "blue",
        xlab="Churn", ylab= i)
}


## Correlation Plot 
 corrgram(data[,numeric_index], order = F,
          upper.panel=panel.pie, text.panel=panel.txt, main = "Correlation Plot")
 
## Chi-squared Test of Independence
 factor_index = sapply(data,is.factor)
 factor_data = data[,factor_index]
 
 for (i in 1:4)
 {
   print(names(factor_data)[i])
   print(chisq.test(table(factor_data$Churn,factor_data[,i])))
 }



 ## Dimension Reduction
 data = subset(data, select = -c(phone.number,account.length,state,area.code))
 
 rmExcept('data')
 
 ##################################Feature Scaling################################################
 #Normality check
 
 #hist(data$number.vmail.messages, col='yellow')
 
 #####Normalization
 
 for(i in cnames){
   print(i)
   data[,i]=(data[,i]-min(data[,i]))/(max(data[,i]-min(data[,i])))
 }
####Partitioning dataset into train and test 
train=data[1:3333,]
test=data[3334:5000,]

######Model Development#############
###Decision Tree#########
C50m=C5.0(Churn~.,train,trials=100,rules=TRUE)

#Summary of Decision Tree model
summary(C50m)

#Predictions for test case
C50_Pred=predict(C50m,test[,-17],type="class")

##Confusion Matrix for D-Tree
ConfM_C50=table(test$Churn,C50_Pred)
confusionMatrix(ConfM_C50)

#Accuracy:96.4%
#FNR:23.66% 


##KNN Implementation
#Import class library to access knn method
library(class)

#Predict test data
KNN_Predict=knn(train[,1:16],test[,1:16],train$Churn,k=7)

#Confusion matrix
ConfM=table(KNN_Predict,test$Churn)
ConfM=table(KNN_Predict)

#Accuracy : 90.16%
sum(diag(ConfM))/nrow(test)

#False Negative rate : 15.11%
#FNR = FN/FN+TP 

######################Random Forest
Rf=randomForest(Churn~.,train,importance=TRUE,ntree=500)

#Extract rules fromn random forest
#transform rf object to an inTrees' format
treeList=RF2List(Rf)
# 
# #Extract rules
 exec=extractRules(treeList,train[,-17])  # R-executable conditions
# 
# #Visualize some rules
 exec[1:2,]
# 
# #Make rules more readable:
 readableRules=presentRules(exec,colnames(train))
 readableRules[1:2,]
# 
# #Get rule metrics
 ruleMetric=getRuleMetric(exec,train[,-17],train$Churn)  # get rule metrics
# 
# #evaulate few rules
 ruleMetric[1:2,]

#Predict test data using random forest model
RF_Predictions=predict(RF_model,test[,-17])

##Evaluate the performance of classification model
ConfMatrix_RF=table(test$Churn,RF_Predictions)
confusionMatrix(ConfMatrix_RF)
confusionMatrix
#False Negative rate
#FNR = FN/FN+TP 

#Accuracy = 96.16%
#FNR = 26.33%

File: ChurnReduction.py

#Load libraries
import os
import pandas as pd
import numpy as np
from fancyimpute import KNN   
import matplotlib.pyplot as plt
from scipy.stats import chi2_contingency

#Set working directory
os.chdir("/Users/anchalsharma/Documents/Edwisor/Project 1")
#Load data
Churn_train = pd.read_csv("Train_data.csv")
Churn_test = pd.read_csv("Test_data.csv")
data=Churn_train.append(Churn_test)


# ## Missing Value Analysis

#Creating a dataframe with missing percentage
missing_val = pd.DataFrame(data.isnull().sum())
missing_val

data['state']=data['state'].astype('object')


#Assigning levels to the categories
lis = []
for i in range(0,data.shape[1]):
    if(data.iloc[:,i].dtypes=='object'):
        data.iloc[:,i]=pd.Categorical(data.iloc[:,i])
        data.iloc[:,i]=data.iloc[:,i].cat.codes 
        data.iloc[:,i]=data.iloc[:,i].astype('object')
        lis.append(data.columns[i])      


# ## Outlier Analysis

# #Plot boxplot to visualize Outliers
#%matplotlib inline  
#plt.boxplot(data['total eve minutes'], notch=True)

#save numeric names
cnames=["number vmail messages","total day minutes",
        "total day calls","total day charge",
        "total eve minutes","total eve calls",
        "total eve charge","total night minutes",
        "total night calls","total night charge",
        "total intl minutes","total intl calls","total intl charge",
        "number customer service calls"]


# ## Feature Selection

data = data.drop(['state','area code','phone number','account length'], axis=1)

##Correlation analysis
#Correlation plot
df = data.loc[:,cnames]

#Generating correlation matrix
corr = df.corr()

#Plot using seaborn library
import seaborn as sns
sns.heatmap(corr,vmin=-1,vmax=1)

#Chisquare test of independence
#Save categorical variables
cat_names = ["international plan", "voice mail plan"]

#loop for chi square values
for i in cat_names:
    print(i)
    chi2,p,dof,ex=chi2_contingency(pd.crosstab(data['Churn'],data[i]))
    print(p)


## Feature Scaling

#dt = data.copy()
#data = dt.copy()

#Normality check
get_ipython().run_line_magic('matplotlib', 'inline')
for i in cnames:
    plt.hist(data[i], bins='auto')

#Nomalisation
for i in cnames:
    print(i)
    data[i]=(data[i]-min(data[i]))/(max(data[i])-min(data[i]))


# ## Model Development

# ## Decision Tree

#Import Library for decision tree
from sklearn import tree

data.head()


# In[ ]:


#replace target categories with True and False
data['Churn']=data['Churn'].replace(0,'False.')
data['Churn']=data['Churn'].replace(1,'True.')

#Divide the data into Train and Test
X_train=data.iloc[0:3333,0:16]
X_test=data.iloc[3333:,0:16]
y_train=data.iloc[0:3333,16]
y_test=data.iloc[3333:,16]

#Decision Tree
C50m = tree.DecisionTreeClassifier(criterion='entropy').fit(X_train, y_train)

#predict test cases
C50_Pred=C50m.predict(X_test)

#build confusion matrix
CM = pd.crosstab(y_test, C50_Pred)

#Assign TP, TN, FP, FN
TN = CM.iloc[0,0]
FN = CM.iloc[1,0]
TP = CM.iloc[1,1]
FP = CM.iloc[0,1]

#check accuracy of model:92
((TP+TN)*100)/(TP+TN+FP+FN)

#False Negative rate:30
#(FN*100)/(FN+TP)

#Results
#Accuracy: 92
#FNR: 30

(FN*100)/(FN+TP)

## Random Forest

#replace target categories with True and False
data['Churn']=data['Churn'].replace(0,'False.')
data['Churn']=data['Churn'].replace(1,'True.')


X_train=data.iloc[0:3333,0:16]
X_test=data.iloc[3333:,0:16]
y_train=data.iloc[0:3333,16]
y_test=data.iloc[3333:,16]

#Random Forest
from sklearn.ensemble import RandomForestClassifier

RF_model = RandomForestClassifier(n_estimators=500).fit(X_train, y_train)

RF_Predictions = RF_model.predict(X_test)

#build confusion matrix
# from sklearn.metrics import confusion_matrix 
# CM = confusion_matrix(y_test, y_pred)
CM = pd.crosstab(y_test, RF_Predictions)

#let us save TP, TN, FP, FN
TN = CM.iloc[0,0]
FN = CM.iloc[1,0]
TP = CM.iloc[1,1]
FP = CM.iloc[0,1]

#check accuracy of model
#accuracy_score(y_test, y_pred)*100
((TP+TN)*100)/(TP+TN+FP+FN)

#False Negative rate :25
#(FN*100)/(FN+TP)

#Accuracy: 95(100 trees)96 (500 trees)
#FNR: 28(100 trees) 25(500 Trees)

(FN*100)/(FN+TP)


## KNN Implementation

data['Churn'] = data['Churn'].replace( 'False.',1)
data['Churn'] = data['Churn'].replace( 'True.',2)

X_train=data.iloc[0:3333,0:16]
X_test=data.iloc[3333:,0:16]
y_train=data.iloc[0:3333,16]
y_test=data.iloc[3333:,16]

from sklearn.neighbors import KNeighborsClassifier

KNN_model=KNeighborsClassifier(n_neighbors=7).fit(X_train, y_train)

#predict test cases
KNN_Predictions=KNN_model.predict(X_test)

#build confusion matrix
CM=pd.crosstab(y_test, KNN_Predictions)

#let us save TP, TN, FP, FN
TN=CM.iloc[0,0]
FN=CM.iloc[1,0]
TP=CM.iloc[1,1]
FP=CM.iloc[0,1]

#check accuracy of model
((TP+TN)*100)/(TP+TN+FP+FN)

#False Negative rate 
#(FN*100)/(FN+TP)

#Accuracy: 88/90/91/91 n_Neighbours_value:(1/3/5/7)
#FNR: 49/51/52/56


## Naive Bayes

data['Churn']=data['Churn'].replace('False.',1)
data['Churn']=data['Churn'].replace('True.',2)

X_train=data.iloc[0:3333,0:16]
X_test=data.iloc[3333:,0:16]
y_train=data.iloc[0:3333,16]
y_test=data.iloc[3333:,16]

from sklearn.naive_bayes import GaussianNB

#Naive Bayes implementation
NB_m=GaussianNB().fit(X_train,y_train)

#predict test cases
NB_Pred=NB_m.predict(X_test)

#Build confusion matrix
CM=pd.crosstab(y_test,NB_Pred)

#let us save TP, TN, FP, FN
TN=CM.iloc[0,0]
FN=CM.iloc[1,0]
TP=CM.iloc[1,1]
FP=CM.iloc[0,1]

#To calclulate accuracy of model
((TP+TN)*100)/(TP+TN+FP+FN)

#False Negative rate 
#(FN*100)/(FN+TP)

#Accuracy: 87
#FNR: 47

